### PR TITLE
Eiffel-ER integration

### PR DESCRIPTION
**Pull request by ewahjon**
Added application property for Eiffel-ER url and settings file name. Vici can now fetch Eiffel events from the current version of Eiffel-ER API.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 # Logging file, will not put out to a file if not defined.
 #logging.file=vici.log
-
 # Logging
 logging.level.root=INFO
-logging.level.org.springframework.web=INFO
\ No newline at end of file
+logging.level.org.springframework.web=INFO
+# Settings
+eiffel.vici.settings-file-name=settings.json
```
Comment (tobiasake) [inline src/main/resources/application.properties]: What is the content of settings.json? Can't allsettings be located in this application.properties file?

File: src/main/java/com/ericsson/vici/api/entities/Settings.java [hunk 2]
```diff
@@ -17,16 +17,26 @@
...
 import static com.ericsson.vici.Fetcher.*;
 
+@Data
+@NoArgsConstructor
 public class Settings {
-    public static final String propertiesVersion = "2.2 - Added cut at events, changed standard values to fit new algorithm.";
 
-    private static final String[] eventTypesAggregatedOn = {
+    @JsonIgnore
+    private final String latestVersion = "2.3 - Eiffel-ER with Spring property as default.";
```
Comment (tobiasake) [inline src/main/java/com/ericsson/vici/api/entities/Settings.java]: Why is this needed or what is it used for?

File: src/main/java/com/ericsson/vici/Fetcher.java [hunk 3]
```diff
@@ -143,60 +146,64 @@ public Events fetchEvents(Preferences preferences) {
...
-                e.printStackTrace();
-            }
+        } else { // Eiffel ER Rest API version??
+            URIBuilder builder;
+            int totalNumberItems = Integer.MAX_VALUE;
+//            int pageSize = 59923; // seems to be the maximum of events the Eiffel ER can handle at the moment.
+            int pageSize = 2000;
+            int page = 1;
+
+            while (eiffelEvents.size() < totalNumberItems) {
```
Comment (evasiba-ericsson) [inline src/main/java/com/ericsson/vici/Fetcher.java]: what happens if you have 50 pages and 100000 events and while you fetch the 5th page 10 events will come in? totalNumberOfItems will be 100010 and the latest 10 will show up in the first page not last. The loop will loop for ever since you will never read the last 10 events that came after reading the first page. the events size will be 10 less than the total number of items that you fetch continuously from each response.

File: src/main/java/com/ericsson/vici/api/SettingsHandler.java [hunk 4]
```diff
@@ -22,25 +22,49 @@
...
-        } else if (settings.getVersion() == null || !settings.getVersion().equals(propertiesVersion)) {
+        if (settings == null || !settings.isUpToDate()) {
             resetSettingsDefault();
         }
     }
 
+    public Settings getDefaultSettings() {
+        Settings settings = new Settings();
+
+        List<EiffelEventRepository> STANDARD_EIFFEL_REPOSITORIES = Arrays.asList(
```
Comment (evasiba-ericsson) [inline src/main/java/com/ericsson/vici/api/SettingsHandler.java]: Use camel case for naming of variables.

File: src/main/resources/application.properties (hunk 1 above)
Comment (ewahjon) [reply src/main/resources/application.properties]: The idea was to let a single instance of Eiffel Vici be able to handle multiple sources (CI-systems?) which may want to have specific settings depending on how the system is build and what data is available.
Maybe this is not needed anymore in that case we should create a new issue about that

File: src/main/java/com/ericsson/vici/api/entities/Settings.java (hunk 2 above)
Comment (ewahjon) [reply src/main/java/com/ericsson/vici/api/entities/Settings.java]: This is used for overriding old settings if a user pulls down a new version with different structured settings (most relevant during development before the structure of settings etc is more stable). The idea is to compare the stored settings whit the latest version to see if a migration of settings is required

File: src/main/java/com/ericsson/vici/Fetcher.java [hunk 5]
```diff
@@ -143,60 +146,64 @@ public Events fetchEvents(Preferences preferences) {
...
-            JSONObject queryJson = null;
-            try {
-                queryJson = new JSONObject(mapper.writeValueAsString(query));
-            } catch (JsonProcessingException e) {
-                e.printStackTrace();
-            }
+        } else { // Eiffel ER Rest API version??
+            URIBuilder builder;
+            int totalNumberItems = Integer.MAX_VALUE;
+//            int pageSize = 59923; // seems to be the maximum of events the Eiffel ER can handle at the moment.
```
[the remark this answers is not in the document]
Comment (ewahjon) [reply src/main/java/com/ericsson/vici/Fetcher.java]: Yes you are right!

File: src/main/resources/application.properties (hunk 1 above)
Comment (tobiasake) [reply src/main/resources/application.properties]: ok

File: src/main/java/com/ericsson/vici/api/entities/Settings.java (hunk 2 above)
Comment (tobiasake) [reply src/main/java/com/ericsson/vici/api/entities/Settings.java]: ok

**Comment by ewahjon** [comment]
Good feedback, will make a new commit where I completely remove the settings feature sicne there seem to be no need for a Vici instance to fetch from multiple repositories, making it much easier to understand the code. Moving more settings to the application.properties.

I will implement more filtered requests to the Eiffel-er giving the user more control of what data to fetch. This is dependent on what is currently possible with the Eiffel-ER API.

File: src/main/java/com/ericsson/vici/Fetcher.java (hunk 3 above)
Comment (ewahjon) [reply src/main/java/com/ericsson/vici/Fetcher.java]: If the Eiffel-ER works as expected that should be handled by the `totalNumberItems = response.totalNumberItems;` which is updated every loop. 
But yes there is a problem here if the number of events changes in the event repository while Vici is fetching the events.
But at least this makes it possible to start testing the eiffel-er with vici for now. A temporary solution. Will be fixed in next commit.

File: src/main/java/com/ericsson/vici/Fetcher.java [hunk 6]
```diff
@@ -143,60 +146,64 @@ public Events fetchEvents(Preferences preferences) {
         log.info("Downloading eiffel-events from: " + preferences.getUrl());
 
         RestTemplate restTemplate = new RestTemplate();
-        ResponseEntity<EiffelEvent[]> responseEntity;
-        EiffelEvent[] eiffelEvents = null;
+        ResponseEntity<EiffelEREventsResponse> responseEntity;
```
[the remark this answers is not in the document]
Comment (ewahjon) [reply src/main/java/com/ericsson/vici/Fetcher.java]: Good points!
